File: shoring/fxapp/src/main/java/com/panopset/fxapp/AnchorFactory.java
package com.panopset.fxapp;

import java.util.ArrayList;
import java.util.List;

public enum AnchorFactory {

  INSTANCE;
  
  public static void add(Anchor anchor) {
    INSTANCE.addAnchor(anchor);
  }

  public static Anchor findAnchor() {
    return INSTANCE.getTopAnchor();
  }
  
  public static List<Anchor> getAnchors() {
    return INSTANCE.getCopyOfAnchors();
  }

  public static void remove(Anchor anchor) {
    INSTANCE.drop(anchor);
  }
 
  private synchronized void addAnchor(Anchor anchor) {
    if (anchors.contains(anchor)) {
      return;
    }
    anchors.add(anchor);
  }

  private synchronized Anchor getTopAnchor() {
    if (anchors.isEmpty()) {
      return null;
    }
    return anchors.get(0);
  }

  private synchronized List<Anchor> getCopyOfAnchors() {
    List<Anchor> rtn = new ArrayList<>();
    for (Anchor anchor : anchors) {
      rtn.add(anchor);
    }
    return rtn;
  }

  private synchronized void drop(Anchor anchor) {
    if (anchors.contains(anchor)) {
      if (anchors.size() == 1) {
        JavaFXapp.panExit();
      } else {
        anchors.remove(anchor);
      }
    }
  }

  private final List<Anchor> anchors = new ArrayList<>();
}

File: code/shoring/fxapp/src/main/kotlin/com/panopset/fxapp/AnchorFactory.kt
package com.panopset.fxapp

object AnchorFactory {

    fun addAnchor(anchor: Anchor) {
        if (anchors.contains(anchor)) {
            return
        }
        anchors.add(anchor)
    }

    private val topAnchor: Anchor?
        get() = if (anchors.isEmpty()) {
            null
        } else anchors[0]

    private val copyOfAnchors: List<Anchor>
        get() {
            val rtn: MutableList<Anchor> = ArrayList()
            for (anchor in anchors) {
                rtn.add(anchor)
            }
            return rtn
        }

    private fun drop(anchor: Anchor) {
        if (anchors.contains(anchor)) {
            if (anchors.size == 1) {
                JavaFXapp.panExit()
            } else {
                anchors.remove(anchor)
            }
        }
    }

    private val anchors: MutableList<Anchor> = ArrayList()

    fun findAnchor(): Anchor? {
        return topAnchor
    }

    fun getAnchors(): List<Anchor> {
        return copyOfAnchors
    }

    fun removeAnchor(anchor: Anchor) {
        drop(anchor)
    }
}
